### PR TITLE
Upgrade to latest security release: go1.20.4 (#9370)

### PR DESCRIPTION
**Pull request by salvacorts**
**What this PR does / why we need it**:
https://groups.google.com/g/golang-announce/c/MEb0UyuSMsU

Also updated the build image.

---------

Signed-off-by: Danny Kopping <danny.kopping@grafana.com>
(cherry picked from commit f5c0fdf02873ec353a7cb4cc4f17c06431412947)

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

File: .drone/drone.yml [hunk 1]
```diff
@@ -93,14 +93,14 @@ steps:
   depends_on:
   - clone
   environment: {}
-  image: grafana/loki-build-image:0.28.1
+  image: grafana/loki-build-image:0.28.2
```
Comment (dannykopping) [inline .drone/drone.yml:96]: This is stale; it should be `0.28.3`

**Review by dannykopping — CHANGES_REQUESTED** [review]

File: .drone/drone.yml (hunk 1 above)
Comment (salvacorts) [reply .drone/drone.yml:96]: That'll be updated on a separate PR where I'll cherry pick https://github.com/grafana/loki/pull/9372

Comment (salvacorts) [reply .drone/drone.yml]: Done in this PR

**Review by dannykopping — APPROVED** [review]
LGTM

**Pull request merged**